cloud-init/qemu_performance.py

```diff
@@ -0,0 +1,549 @@
+                        service, orig_avg[key][service], new_avg[key][service]
+                    )
+                    if perf_delta:
+                        perf_deltas.append(PERF_LINE.format(**perf_delta))
+    if perf_deltas:
+        perf_deltas.insert(0, PERF_HEADER)
+    for delta in perf_deltas:
+        print(delta)
+    if not perf_deltas:
+        print("------ No significant boot speed deltas seen -----")
+    log_avg_table("Control", orig_avg)
+    log_avg_table("Updated cloud-init", new_avg)
+
+
+def inspect_boot_time_deltas_for_deb(
+    deb_path: str,
+    sample_count: int,
+    release: str,
+    data_dir: Path,
+    user_data: str,
+):
+    """
+    Download daily image for a given release.
+    Boot that daily image and inspect typical data related
+    to boottime:
+     1. systemd-analyze blame:
+        - to see costs of cloud-init units vs snapd.seeded.service
+     2. systemd-analyze critical-chain systemd-user-sessions.service
+        - generally blocks console login and non-root SSH access to VM
+     3. systemd-analyze critical-chain cloud-config.service
+        - which typically blocks on a costly snapd.seeded.service for this VM
+     4. systemd-analyze critical-chain snapd.seeded.service:
+        - to ensure we are seeding snaps during this boot and didn't use a
+          dirty image which already seeded
+     5. cloud-init analyze blame: To highlight where time is spent.
+        Particular attention to cc_apt_configure additional time cost
+     6. grep update /var/log/cloud-init.log to assert whether apt-get update
+        is called in default cases.
+
+    Use mount-image-callback to inject deb_path into a derivative daily image.
+    Boot original daily image and report boottime stats.
+    Boot derivative image and report boottime stats.
+    """
+    orig_data = []
+    new_data = []
+    existing_data_files = [f for f in data_dir.rglob("*.json")]
+    if existing_data_files:
+        print(f"--- Using pre-existing data files in {data_dir}")
+        daily = "pre-existing samples"
+        for f in existing_data_files:
+            if "avg" in f.name:
+                continue
+            if "orig" in f.name:
+                orig_data.append(json.loads(f.read_text()))
+            else:
+                new_data.append(json.loads(f.read_text()))
+    else:
+        data_dir.mkdir(exist_ok=True)
+        with pycloudlib.Qemu(tag="examples") as cloud:
+            daily = cloud.daily_image(release=release)
+            print(
+                f"--- Creating modified daily image {daily} with cloud-init"
+                f" from {deb_path}"
+            )
+            new_image = update_cloud_init_in_img(daily, deb_path, suffix=1)
+            print(
+                f"--- Launching {sample_count} control daily images {daily} ---"
+            )
+            for sample in range(sample_count):
+                instance = cloud.launch(image_id=daily, user_data=user_data)
```

Comment:
For large `sample_count` launches, it would be convenient to have a status output counter.